### PR TITLE
drivers: gpio: Xilinx Processor System MIO / EMIO GPIO controller driver

### PR DESCRIPTION
**Pull request by ibirnbaum**
Driver implementation plus device tree and SoC initialization code (MMU setup) integration for the Xilinx Processor System MIO / EMIO GPIO controller as contained in both the Zynq-7000 and the ZynqMP (UltraScale+) SoC families.

No modification of the ZynqMP initialization code (MPU regions setup) is required, as the controller's register space is already covered by an existing region definition.

### CODE REVIEW

**Review by tejlmand — COMMENTED** [review]
build system parts lgtm.

**Review by mbolivar-nordic — COMMENTED** [review]
A bit of a strange GPIO driver, but no  objections from me on the DT portions. Just a few minor things.

(I don't get the 'EOF' comments, but I'm not blocking on those...)

File: drivers/gpio/gpio_xlnx_ps.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	}
+}
+
+/*
+ * Register & initialize all instances of the Processor System's MIO / EMIO GPIO
+ * controller specified in the device tree.
+ */
+DT_INST_FOREACH_STATUS_OKAY(GPIO_XLNX_PS_DEV_INITITALIZE);
+
+/* EOF */
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps.c]: How about "That's all, folks!" instead? ;-)
Having end-of-file markers is just this habit that goes way back. I've now seen that those are neither common nor required around here, so for this driver, I've removed them. I'll later remove them from my other sources as well.

File: dts/bindings/gpio/xlnx,ps-gpio-bank.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#
+# Copyright (c) 2022, Weidmueller Interface GmbH & Co. KG
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: |
+  Xilinx Zynq-7000/ZynqMP MIO/EMIO GPIO controller bank node.
+
+  Each node of this type specified in the device tree represents
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply dts/bindings/gpio/xlnx,ps-gpio-bank.yaml]: Noted & changed accordingly.

File: dts/bindings/gpio/xlnx,ps-gpio-bank.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  a bank of the MIO/EMIO GPIO controller integrated in the Processor
+  System.
+
+compatible: "xlnx,ps-gpio-bank"
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply dts/bindings/gpio/xlnx,ps-gpio-bank.yaml:13]: Noted for future driver implementations.

**Comment by ibirnbaum** [comment]
> A bit of a strange GPIO driver

I guess GPIO controllers which first provide you with a large number of I/O pins, but then require you to basically hardwire them & have the configuration tied in part to the FPGA bitstream are uncommon. Also, maxing out the MIO (external) GPIO pins will most likely result in you having very little unused I/O pin resources left in particular in the smaller footprints of the Zynq-7000 series. At least for this controller, interrupt handling is straightforward, while Xilinx's other supported GPIO driver, the IP-core-based AXI GPIO controller, does have an interrupt line and an interrupt pending bit, but no per-pin interrupt status bits. Therefore, if you want interrupt support for that controller, every value change upon data register read and write operations must be stored so that it can be determined within the ISR which pin actually triggered the interrupt. So all in all, the design isn't that bad, just the large number of pins requires this bank-wise access workaround due to bit mask size limitations.

Thanks for the review, Marti!

**Review by mbolivar-nordic — DISMISSED** [review]
@mnkp please review

File: drivers/gpio/gpio_xlnx_ps.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * the callbacks registered for the respective bank are triggered
+ * using the functionality provided by the GPIO sub-system.
+ *
+ * @param dev Pointer to the PS GPIO controller's device.
+ */
+static void gpio_xlnx_ps_isr(const struct device *dev)
+{
+	const struct gpio_xlnx_ps_dev_cfg *dev_conf = dev->config;
+
+	struct gpio_driver_api *api;
```
Comment (stephanosio) [inline drivers/gpio/gpio_xlnx_ps.c:68]:
```suggestion
	const struct gpio_driver_api *api;
```

File: drivers/gpio/gpio_xlnx_ps.c [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	struct gpio_xlnx_ps_bank_dev_data *bank_data;
+
+	uint32_t bank;
+	uint32_t int_mask;
+
+	for (bank = 0; bank < dev_conf->num_banks; bank++) {
+		api = (struct gpio_driver_api *)dev_conf->bank_devices[bank]->api;
```
Comment (stephanosio) [inline drivers/gpio/gpio_xlnx_ps.c:75]:
```suggestion
		api = dev_conf->bank_devices[bank]->api;
```

cast no longer necessary when `api` is declared const.

File: drivers/gpio/gpio_xlnx_ps.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+/*
+ * Macro used to generate each parent controller device's IRQ attach
+ * function.
+ */
+#define GPIO_XLNX_PS_DEV_CONFIG_IRQ_FUNC(idx)\
+static void gpio_xlnx_ps##idx##_irq_config(const struct device *dev)\
+{\
+	ARG_UNUSED(dev);\
+	IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority),\
+	gpio_xlnx_ps_isr, DEVICE_DT_INST_GET(idx), 0);\
```
Comment (stephanosio) [inline drivers/gpio/gpio_xlnx_ps.h:48]: Nit: alignment

```suggestion
	IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority),\
		    gpio_xlnx_ps_isr, DEVICE_DT_INST_GET(idx), 0);\
```

File: drivers/gpio/gpio_xlnx_ps.c (hunk 4 above)
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps.c]: added const qualifier as suggested.

File: drivers/gpio/gpio_xlnx_ps.c (hunk 5 above)
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps.c]: removed type cast as suggested.

File: drivers/gpio/gpio_xlnx_ps.h (hunk 6 above)
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps.h]: modified indentation as suggested.

**Review by stephanosio — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
I'm suggesting a DNM until @mnkp can review as GPIO maintainer.

**Review by mnkp — CHANGES_REQUESTED** [review]
Could you check if the driver is passing `tests/drivers/gpio/gpio_api_1pin/` test. And possibly also `tests/drivers/gpio/gpio_basic_api/`.

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+		int_polarity_data	&= ~pin_mask;
+		int_any_data		&= ~pin_mask;
+	}
+
+	sys_write32(int_type_data,     GPIO_XLNX_PS_BANK_INT_TYPE_REG);
+	sys_write32(int_polarity_data, GPIO_XLNX_PS_BANK_INT_POLARITY_REG);
+	sys_write32(int_any_data,      GPIO_XLNX_PS_BANK_INT_ANY_REG);
+
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		sys_write32(pin_mask, GPIO_XLNX_PS_BANK_INT_EN_REG);
```
Comment (mnkp) [inline drivers/gpio/gpio_xlnx_ps_bank.c:331]: I'm not familiar with the architecture but it seems we should write to `GPIO_XLNX_PS_BANK_INT_STAT_REG` to clear possibly dirty interrupt status bit before enabling the interrupt.

File: drivers/gpio/gpio_xlnx_ps.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+ * Top-level device initialization macro, executed for each PS GPIO
+ * parent device entry contained in the device tree which has status
+ * "okay".
+ */
+#define GPIO_XLNX_PS_DEV_INITITALIZE(idx)\
+GPIO_XLNX_PS_GEN_BANK_ARRAY(idx)\
+GPIO_XLNX_PS_DEV_CONFIG_IRQ_FUNC(idx)\
+GPIO_XLNX_PS_DEV_DATA(idx)\
+GPIO_XLNX_PS_DEV_CONFIG(idx)\
+GPIO_XLNX_PS_DEV_DEFINE(idx)
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps.h]: Moved the macros to the .c file.

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+ */
+static int gpio_xlnx_ps_pin_configure(const struct device *dev,
+				      gpio_pin_t pin,
+				      gpio_flags_t flags)
+{
+	const struct gpio_xlnx_ps_bank_dev_cfg	*dev_conf = dev->config;
+	uint32_t				pin_mask  = BIT(pin);
+	uint32_t				bank_data;
+	uint32_t				dirm_data;
+	uint32_t				oen_data;
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c]: Changed the formatting.

File: drivers/gpio/gpio_xlnx_ps_bank.h [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+#define GPIO_XLNX_PS_BANK_INT_POLARITY_REG  ((dev_conf->base_addr + 0x220)\
+	+ ((uint32_t)dev_conf->bank_index * 0x40))
+#define GPIO_XLNX_PS_BANK_INT_ANY_REG       ((dev_conf->base_addr + 0x224)\
+	+ ((uint32_t)dev_conf->bank_index * 0x40))
+
+/* MIO / EMIO bank device definition macros */
+#define GPIO_XLNX_PS_BANK_INIT(idx)\
+static const struct gpio_xlnx_ps_bank_dev_cfg gpio_xlnx_ps_bank##idx##_cfg = {\
+	.common = {\
+		.port_pin_mask = 0xFF,\
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.h]: done.

File: drivers/gpio/gpio_xlnx_ps_bank.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ * which is required for proper operation (such as base memory
+ * addresses etc.) which don't have to and therefore cannot be
+ * modified at run-time.
+ */
+struct gpio_xlnx_ps_bank_dev_cfg {
+	struct gpio_driver_config common;
+
+	uint32_t		  base_addr;
+	uint8_t			  bank_index;
+	uint8_t			  pin_count;
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.h]: ... aaaand it's gone!

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+
+	/*
+	 * Pin index range check -> usually, the last bank of either
+	 * GPIO pin type (MIO or EMIO) has less than the full 32 pins
+	 * available. The pin count for each bank is obtained from the
+	 * device tree.
+	 */
+	if (pin >= dev_conf->pin_count) {
+		return -EINVAL;
+	}
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c]: gone.

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+	/* Check for config flags not supported by the controller */
+	if (flags & (GPIO_PULL_UP | GPIO_PULL_DOWN |
+		     GPIO_OPEN_DRAIN | GPIO_OPEN_SOURCE)) {
+		return -ENOTSUP;
+	}
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c]: Merged  OPEN_DRAIN and OPEN_SOURCE into SINGLE_ENDED as suggested.

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+ *         -EINVAL if the pointer to the output variable is NULL.
+ */
+static int gpio_xlnx_ps_bank_get(const struct device *dev,
+				 gpio_port_value_t *value)
+{
+	const struct gpio_xlnx_ps_bank_dev_cfg *dev_conf = dev->config;
+
+	if (value == NULL) {
+		return -EINVAL;
+	}
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c]: gone.

File: drivers/gpio/gpio_xlnx_ps_bank.c [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+
+	sys_write32(int_type_data,     GPIO_XLNX_PS_BANK_INT_TYPE_REG);
+	sys_write32(int_polarity_data, GPIO_XLNX_PS_BANK_INT_POLARITY_REG);
+	sys_write32(int_any_data,      GPIO_XLNX_PS_BANK_INT_ANY_REG);
+
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		sys_write32(pin_mask, GPIO_XLNX_PS_BANK_INT_EN_REG);
+	} else { /* mode == GPIO_INT_MODE_DISABLED */
+		sys_write32(pin_mask, GPIO_XLNX_PS_BANK_INT_DIS_REG);
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c]: Interrupts are now unconditionally disabled for the respective pin upon entering the config function before any other modifications to the pin's interrupt behaviour are made.

File: drivers/gpio/gpio_xlnx_ps_bank.c (hunk 7 above)
Comment (ibirnbaum) [reply drivers/gpio/gpio_xlnx_ps_bank.c:335]: Implemented as suggested, clearing potential stale pending interrupts makes sense.

**Comment by ibirnbaum** [comment]
Hi Piotr, thanks for the review.

> Could you check if the driver is passing `tests/drivers/gpio/gpio_api_1pin/` test. And possibly also `tests/drivers/gpio/gpio_basic_api/`.

With the changes as committed today, both of those test cases pass:

- gpio_basic_api passes upon addition of an overlay for the board I'm working with (and which I'm going to contribute once this driver is mainlined, I didn't yet want to contribute a board that can't do the blinky sample, it's the Digilent/Avnet Zedboard). There are two GPIO-capable pins available on one of the board's expansion connectors that can be bridged.

- gpio_api_1pin passes, but there's a bit of a board-specific issue here. This test case uses the same pin alias as the blinky sample. Did the board vendor provide an LED tied to a GPIO controlled by this driver? Yes, they did. Does the blink sample work with it? Yes, it does. So does this test case which uses the same DT alias to grab a GPIO pin for testing work? No, it doesn't.

Here's why: the designers of the Zedboard made a bit of an unfortunate design decision. The one LED connected to this GPIO controller, and therefore the obvious choice for the blinky-compatible alias, is connected to pin MIO7, which equals bank 0, pin 7. They could have gone for a few other pins which by default are available for GPIO, such as MIO12 and MIO13, which I used in the overlay for the gpio_basic_api test, or MIO0, which I eventually used for this test. Those would have worked for the gpio_api_1pin test just fine. So why won't MIO7 which the test case accesses via the led0 alias work? The answer is in the Zynq-7000's Technical Reference Manual, section 14.2.3: Out of all the pins managed by this controller, MIO7 and MIO8 are special, they are output-only pins which are used by the SoC to transport voltage control information during the power-on sequence. For exactly those two pins, not even software-generated interrupts work. There seems to be no feedback from the output data to the interrupt detection logic.

Long story short, due to this issue, I changed the led0 alias over from MIO7 to MIO0, and everything worked just fine. Unfortunately, there's no LED on that pin, and I guess having a working mapping for blinky is more important. So anyone wanting to run this test case on a Zedboard will likely have to manually move led0 over to another MIO pin in order to successfully run this test. The pins used for the gpio_basic_api test are always available.

**Review by mnkp — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by ibirnbaum** [comment]
@mbolivar-nordic Now that this has been merged, should the DNM label be removed for completeness' sake?
